### PR TITLE
Make the annotation parser aware about _Unset

### PR DESCRIPTION
**Pull request by sarahhenkens**
## Description

The PR makes the annotation parser aware of the `_Unset` class type inside the Union with None to allow for strict argument typing and still correctly work with strawberry's type generation.

### Why is this change needed

When specifying a union field arg like this the following:

```python
name: Union[_Unset, Optional[str]] = UNSET,
age: Union[int, None, _Unset] = UNSET,
```

Strawberry will incorrectly try to convert `Union[_Unset, str]` into a GraphQL enum. Preventing the usage of stricter type annotations to ensure type checkers are aware of the third `_Unset` type that is set by defaulting to `UNSET`

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

Relates to #872.

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Sarah Henkens for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/arguments.py
Comment (sarahhenkens) [inline strawberry/arguments.py:33]: I hade to move this to its own file because of a circular dependency

**Comment by BryceBeagle** [comment]
Why are you adding `_Unset` to the `Union`? We set `UNSET` as `Any` (i.e. `UNSET: Any = _Unset()`), so should be able to use it for any assignment without adding it to the annotations, even with `--strict` enabled.

```python
from typing import Any

class _Unset:
    ...
        
UNSET: Any = _Unset()

foo: int = UNSET
```
```
Success: no issues found in 1 source file
```

https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=179e8124da44a9949b94b5b1ec46a4c2

File: tests/schema/test_arguments.py
Comment (BryceBeagle) [inline tests/schema/test_arguments.py:108]: Can these test be narrowed in scope and moved to be with the other type resolving tests for annotations? I don't think we need to going all the way to query execution to test this.

File: tests/schema/test_arguments.py
Comment (BryceBeagle) [inline tests/schema/test_arguments.py:7]: Not sure I like exporting a ~private symbol like this. @patrick91 should we just rename it `Unset`?

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: tests/schema/test_arguments.py
Comment (patrick91) [reply tests/schema/test_arguments.py:7]: I'd say yes. But I wonder if we could find a way to make `UNSET`	work as a type, this:

`x: Unset | SomethingElse = UNSET`

feels a bit strange, what do you think @BryceBeagle @sarahhenkens ?

Comment (sarahhenkens) [reply tests/schema/test_arguments.py:7]: Having `UNSET` being set to `any` is the root cause of failures in type checking to occur. Imagine the following scenario:

```python
@strawberry.field
async def jobs(
    self,
    status: Union[_Unset, Optional[Status]] = UNSET,
    first: Optional[int] = UNSET,
    after: Optional[str] = UNSET,
    last: Optional[int] = UNSET,
    before: Optional[str] = UNSET,
) -> Connection[types.Job]:
    result = await Container.job_service().get_paginated(
        status=status, first=first, after=after, last=last, before=before
    )
 ```

From a type checking perspective `UNSET` is technically not valid value for `first: Optional[int] = UNSET,` because its an instance of `_Unset`, but because `UNSET` has its type overwritten to `any`, type checkers will just ignore it.

This gives code a false sense of security, this is why my code example for `status` has an explicit type for it to ensure downstream code is aware of it.

If `Unset` has its correct type set to `_Unset`, type checkers will prevent you from doing  `first: Optional[int] = UNSET,` which is in my opinion a much better default behavior than having `UNSET` be an `any`.

I wanted to keep backward compatibility since other projects might assume it to be any so I just made changes to the internals to support this special type without making changes to the public API and its types.

Comment (sarahhenkens) [reply tests/schema/test_arguments.py:7]: > I'd say yes. But I wonder if we could find a way to make UNSET work as a type, this:
> x: Unset | SomethingElse = UNSET
> feels a bit strange, what do you think @BryceBeagle @sarahhenkens ?

At the core, `x: Optional[T] = UNSET` is essential an 3-way value, "null, not-set, T". In my original ticket (#872) I recommended a new type to represent this tri-state:

```python
T = TypeVar("T")
UnsetOrOptional = Union[_Unset, Optional[T]]
```

This would make the interface much cleaner:
```python
status: UnsetOrOptional[Status]] = UNSET,
```

File: tests/schema/test_arguments.py
Comment (sarahhenkens) [reply tests/schema/test_arguments.py:108]: Great call out!!

I choose this test since it's the closest to an integration test that ensures the annotation parser works and it correctly still considers the schema output to be a nullable field.

I've updated the tests in `tests/types/resolving/test_optionals.py` as well. Should I remove the ones in `test_arguments`? Or would it be useful to keep them in here as well?

File: tests/schema/test_arguments.py
Comment (patrick91) [reply tests/schema/test_arguments.py:7]: I think I was suggesting to be able to do this:

```
x: UNSET | SomethingElse = UNSET
```

treating UNSET almost like a literal, so wouldn't need to expose `_Unset` ^^

But I'm not sure this is easy to do at the moment!

Comment (sarahhenkens) [reply tests/schema/test_arguments.py:7]: Having a clearer API sound great! 

My changes are just in the core logic to not have it break on its own `_Unset` though.

Thoughts on getting this merged to unblock any projects that want the strict typing while we work in the future on a better general `UNSET` behavior? 

I wanted to keep this PR's scope minimal by just fixing the annotation parser bug.

Comment (patrick91) [reply tests/schema/test_arguments.py:7]: sounds sensible! I'll open an issue for the (potential) improvement :)

**Review by patrick91 — APPROVED** [review]
@sarahhenkens looks good to me! Are you ok with keeping using `_Unset` for now, or do you want to rename it to `Unset`?

For me we can keep `_Unset` until we deal with https://github.com/strawberry-graphql/strawberry/issues/1473 so it is clear that this class is not really meant to be used :)

**Comment by sarahhenkens** [comment]
@patrick91 I think we should keep it as `_Unset` to be 100% we do not break compatibility for projects that might be using it.

File: tests/schema/test_arguments.py
Comment (sarahhenkens) [reply tests/schema/test_arguments.py:108]: @BryceBeagle can you take another 👀  at this PR?

Comment (BryceBeagle) [reply tests/schema/test_arguments.py:108]: With the tests in tests/types/resolving I think these tests here are redundant. No codepath should escape those tests and make it here undetected.

File: tests/types/resolving/test_optionals.py
Comment (BryceBeagle) [inline tests/types/resolving/test_optionals.py:31]: Can we add a test like this for normal `Union`s (without `None`; e.g. `Union[SomeType, OtherType, _Unset]`)?

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: tests/types/resolving/test_optionals.py
Comment (sarahhenkens) [reply tests/types/resolving/test_optionals.py:31]: Great idea!!

I was writing a test case for this but then I realized that if I specify a `Union[str, int, _Unset]`, it does not get stripped out (existing behavior) because the change in this PR is only occurring in the code path that hits optionals.

Optional types are detected by strawberry when it's a union that contains a `None`. We would need to add a `None` in this test case which doesn't make sense to add it to the union tests.

Instead, I've added another assertion into the `test_optionals`:

`test_optional_union_containing_a_real_union_and_unset`

This asserts that `Union[_Unset, None, Foo, Bar]` is resolved as a proper optional `Union[Foo, Bar]` without a default.

File: tests/schema/test_arguments.py
Comment (sarahhenkens) [reply tests/schema/test_arguments.py]: Updated!

File: tests/types/resolving/test_optionals.py
Comment (BryceBeagle) [reply tests/types/resolving/test_optionals.py:31]: Hmm that makes me wonder if this logic should instead be moved (or mirrored) to to `.create_union`.

Also, what happens if someone does `Optional[_Unset]` right now? I think that should probably throw an exception.

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: tests/types/resolving/test_optionals.py
Comment (BryceBeagle) [reply tests/types/resolving/test_optionals.py:31]: Hmm that makes me wonder if this logic should instead be moved (or mirrored) to to `.create_union`.

Also, what happens if someone does `Optional[_Unset]` right now? I think that should probably throw an exception.

Comment (sarahhenkens) [reply tests/types/resolving/test_optionals.py:31]: IMO that is too much of an edge case to handle here since `_Unset` is already by design an internal type.

Comment (BryceBeagle) [reply tests/types/resolving/test_optionals.py:31]: Hmm, @patrick91 what do you think?

Comment (patrick91) [reply tests/types/resolving/test_optionals.py:31]: sorry I missed this comment, I think it is fine to leave it! we can deal with this case when working on https://github.com/strawberry-graphql/strawberry/issues/1473

**Review by BryceBeagle — APPROVED** [review]

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**
